tedeous/callbacks/plot.py

```diff
@@ -39,43 +48,270 @@ def _print_nn(self):
+                    ax1 = fig.add_subplot(1, self.nvars_model, i + 1)
+                    if self.title is not None:
+                        ax1.set_title(self.title + ' variable {}'.format(i))
+                    ax1.scatter(self.grid.detach().cpu().numpy().reshape(-1),
+                                self.net(self.grid)[:, i].detach().cpu().numpy(),
+                                marker='o', s=5, alpha=0.7)
+                else:
+                    ax1 = fig.add_subplot(1, self.nvars_model, i + 1)
+                    if self.title is not None:
+                        ax1.set_title(self.title + f' variable {i}')
+
+                    grid_x = self.grid[:, 0].detach().cpu().numpy()
+                    grid_y = self.grid[:, 1].detach().cpu().numpy()
+                    z_values = self.net(self.grid)[:, i].detach().cpu().numpy()
+
+                    scatter = ax1.scatter(
+                        grid_x, grid_y,
+                        c=z_values,
+                        cmap=cm.jet,
+                        s=5,
+                        alpha=0.7
+                    )
+
+                    ax1.set_xlabel("x1")
+                    ax1.set_ylabel("x2")
+                    fig.colorbar(scatter, ax=ax1)
+
+    def _plot_solution_2d(self, coord_3_values: List[float]):
+        """
+        Solution plot for 2D PDE u(x, y, t) at multiple time points.
+
+        Args:
+            t_values (List[float]): List of time points to plot the solution for.
+        """
+
+        grid_x = self.grid[:, 0].detach().cpu().numpy()
+        grid_y = self.grid[:, 1].detach().cpu().numpy()
+
+        attributes = {'model': ['out_features', 'output_dim', 'width_out'],
+                      'layers': ['out_features', 'output_dim', 'width_out']}
+
+        self.nvars_model = None
+
+        for key, values in attributes.items():
+            for value in values:
+                try:
+                    self.nvars_model = getattr(getattr(self.net, key)[-1], value)
+                    break
+                except AttributeError:
+                    pass
+
+        if self.nvars_model is None:
+            try:
+                self.nvars_model = self.net[-1].out_features
+            except:
+                self.nvars_model = self.net.width_out[-1]
+
+        if self.img_dim == '3d':
+            for coord_3 in coord_3_values:
+                coord_3_mask = self.grid[:, 2] == coord_3
+                fig, axes = plt.subplots(1, self.nvars_model,
+                                         figsize=(8 * self.nvars_model, 5 * self.nvars_model),
+                                         subplot_kw={'projection': '3d'})
+
+                if self.nvars_model == 1:
+                    axes = [axes]
+
+                for i in range(self.nvars_model):
+                    z_values = self.net(self.grid[coord_3_mask])[:, i].detach().cpu().numpy().flatten()
+
+                    axes[i].plot_trisurf(grid_x[coord_3_mask], grid_y[coord_3_mask], z_values,
```

Comment:
example_navier_stokes_2d_TG_vortex.py throws error here:

>TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

We should handle check device stuff here thoughout